stl/inc/__msvc_int128.hpp

```diff
@@ -0,0 +1,1155 @@
+    }
+
+    constexpr _Uint128& operator&=(const _Uint128& _That) noexcept {
+        _Word[0] &= _That._Word[0];
+        _Word[1] &= _That._Word[1];
+        return *this;
+    }
+    template <integral _Ty>
+    friend constexpr _Ty& operator&=(_Ty& _Left, const _Uint128& _Right) noexcept {
+        _Left = static_cast<_Ty>(_Uint128{_Left} & _Right);
+        return _Left;
+    }
+
+    _NODISCARD friend constexpr _Uint128 operator^(const _Uint128& _Left, const _Uint128& _Right) noexcept {
+        return _Uint128{_Left._Word[0] ^ _Right._Word[0], _Left._Word[1] ^ _Right._Word[1]};
+    }
+
+    constexpr _Uint128& operator^=(const _Uint128& _That) noexcept {
+        _Word[0] ^= _That._Word[0];
+        _Word[1] ^= _That._Word[1];
+        return *this;
+    }
+    template <integral _Ty>
+    friend constexpr _Ty& operator^=(_Ty& _Left, const _Uint128& _Right) noexcept {
+        _Left = static_cast<_Ty>(_Uint128{_Left} ^ _Right);
+        return _Left;
+    }
+
+    _NODISCARD friend constexpr _Uint128 operator|(const _Uint128& _Left, const _Uint128& _Right) noexcept {
+        return _Uint128{_Left._Word[0] | _Right._Word[0], _Left._Word[1] | _Right._Word[1]};
+    }
+
+    constexpr _Uint128& operator|=(const _Uint128& _That) noexcept {
+        _Word[0] |= _That._Word[0];
+        _Word[1] |= _That._Word[1];
+        return *this;
+    }
+    template <integral _Ty>
+    friend constexpr _Ty& operator|=(_Ty& _Left, const _Uint128& _Right) noexcept {
+        _Left = static_cast<_Ty>(_Uint128{_Left} | _Right);
+        return _Left;
+    }
+};
+
+template <class>
+class numeric_limits;
+template <>
+class numeric_limits<_Uint128> : public _Num_int_base {
+public:
+    _NODISCARD static constexpr _Uint128(min)() noexcept {
+        return 0;
+    }
+
+    _NODISCARD static constexpr _Uint128(max)() noexcept {
+        return _Uint128{~0ull, ~0ull};
+    }
+
+    _NODISCARD static constexpr _Uint128 lowest() noexcept {
+        return (min) ();
+    }
+
+    _NODISCARD static constexpr _Uint128 epsilon() noexcept {
+        return 0;
+    }
+
+    _NODISCARD static constexpr _Uint128 round_error() noexcept {
+        return 0;
+    }
+
+    _NODISCARD static constexpr _Uint128 denorm_min() noexcept {
+        return 0;
+    }
+
+    _NODISCARD static constexpr _Uint128 infinity() noexcept {
+        return 0;
+    }
+
+    _NODISCARD static constexpr _Uint128 quiet_NaN() noexcept {
+        return 0;
+    }
+
+    _NODISCARD static constexpr _Uint128 signaling_NaN() noexcept {
+        return 0;
+    }
+
+    static constexpr bool is_modulo = true;
+    static constexpr int digits     = 128;
+    static constexpr int digits10   = 38;
+};
+
+template <class...>
+struct common_type;
+template <integral _Ty>
+struct common_type<_Ty, _Uint128> {
+    using type = _Uint128;
+};
+template <integral _Ty>
+struct common_type<_Uint128, _Ty> {
+    using type = _Uint128;
+};
+
+struct _Int128 : _Uint128 {

```

Comment:
Is public inheritance desirable? This means that `_Uint128&` can bind to `_Int128`, but `unsigned int&` can't bind to `int`.

Reply:
This is currently "powering" the cross-type operations. the conversion from unsigned-to-signed is a UDC, whereas the conversion from signed-to-unsigned is a standard conversion, so we unambiguously select `_Uint128` ops. This also makes `true ? _Int128{} : _Uint128{}` work.

But ick no, we don't want mutable references binding like that. I'm going to address this comment last since I think it will be a large time sink.

Reply:
This reminds me of how STL modifiable iterators derive from constant iterators, which was similarly convenient but with hidden harmful effects.